### PR TITLE
opnode: Record errors from batch derivation

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
Errors were not being logged. Note that we keep processing in the
presence of batch derivation errors as we do not trust this L1 input
to always be valid.

**Metadata**
- Fixes ENG-2151

### CODE REVIEW

File: op-node/rollup/derive/payload_attributes.go [hunk 1]
```diff
@@ -224,32 +224,35 @@ func UserDeposits(receipts []*types.Receipt, depositContractAddr common.Address)
...
 				}
 				batches, err := DecodeBatches(config, bytes.NewReader(tx.Data()))
 				if err != nil {
-					// TODO: log/record metric
+					errs = append(errs, fmt.Errorf("invalid batch: tx list: %d, tx: %d, err: %w", i, j, err))
 					continue
 				}
 				out = append(out, batches...)
+			} else {
+				errs = append(errs, fmt.Errorf("invalid to address: tx list: %d, tx: %d", i, j))
```
Comment (tynes) [inline op-node/rollup/derive/payload_attributes.go:251]: This reads weird, maybe:

`invalid "to" address`

or

`invalid target`

File: op-node/rollup/driver/step.go [hunk 2]
```diff
@@ -198,9 +198,14 @@ func (d *outputImpl) insertEpoch(ctx context.Context, l2Head eth.L2BlockRef, l2S
 	if err != nil {
 		return l2Head, l2SafeHead, false, fmt.Errorf("failed to fetch transactions from %s: %v", l1Input, err)
 	}
-	batches, err := derive.BatchesFromEVMTransactions(&d.Config, transactions)
-	if err != nil {
-		return l2Head, l2SafeHead, false, fmt.Errorf("failed to fetch create batches from transactions: %w", err)
+	batches, errs := derive.BatchesFromEVMTransactions(&d.Config, transactions)
```
Comment (tynes) [inline op-node/rollup/driver/step.go:201]: This receives all transactions in a block right? If so, then an invalid `tx.to` isn't an error

Comment (tynes) [inline op-node/rollup/driver/step.go:201]: This receives all transactions in a block right? If so, then an invalid `tx.to` isn't an error

Comment (trianglesphere) [reply op-node/rollup/driver/step.go:201]: Yep, you're right. I'll remove that error

File: op-node/rollup/driver/step.go [hunk 3]
```diff
@@ -198,9 +198,14 @@ func (d *outputImpl) insertEpoch(ctx context.Context, l2Head eth.L2BlockRef, l2S
...
+	// Some input to derive.BatchesFromEVMTransactions may be invalid and produce errors.
+	// We log the errors, but keep going as this process is designed to be resilient to these errors
+	// and we have defaults in case no valid (or partial) batches were submitted.
+	if len(errs) != 0 {
+		for i, err := range errs {
+			d.log.Error("Failed to decode batch", "err_idx", i, "err", err)
+		}
 	}
```
Comment (protolambda) [inline op-node/rollup/driver/step.go:209]: you can just loop over it regardless of length. If `len(errs) == 0` then it's not logging anything anyway.

**Review by protolambda — COMMENTED** [review]
looks ok, you may want to remove the unnecessary if.

**Review by tynes — APPROVED** [review]
Looks good to me, not sure if an invalid sender is technically an error case, we could be able to skip that one. We do want to know about invalid batches for sure

File: op-node/rollup/derive/payload_attributes.go [hunk 4]
```diff
@@ -224,32 +224,33 @@ func UserDeposits(receipts []*types.Receipt, depositContractAddr common.Address)
...
 				seqDataSubmitter, err := l1Signer.Sender(tx) // optimization: only derive sender if To is correct
 				if err != nil {
-					// TODO: log error
+					errs = append(errs, fmt.Errorf("invalid signature: tx list: %d, tx: %d, err: %w", i, j, err))
 					continue // bad signature, ignore
 				}
 				// some random L1 user might have sent a transaction to our batch inbox, ignore them
 				if seqDataSubmitter != config.BatchSenderAddress {
-					// TODO: log/record metric
+					errs = append(errs, fmt.Errorf("unauthorized batch submitter: tx list: %d, tx: %d", i, j))
```
[the remark this answers is not in the document]
Comment (trianglesphere) [reply op-node/rollup/derive/payload_attributes.go:241]: True. It does cost real world money to them & less money to us.

**Review by mslipper — APPROVED** [review]

File: op-node/rollup/derive/payload_attributes.go (hunk 1 above)
Comment (trianglesphere) [reply op-node/rollup/derive/payload_attributes.go]: removed error

**Pull request merged**